### PR TITLE
Build py-numpy with MKL

### PR DESCRIPTION
**Pull request by glennpj**
This PR gets a GCC built python to use MKL for blas/lapack. This is not perfect and I
have put comments for TODO items.

The changes to intel-parallel-studio are to get the correct library path for blas. I was
using that to attempt to use it but there were a couple of other issues. The changes to
the mkl package are to get the proper paths in the environment module, similar to what
was done previously for the intel-parallel-studio package.

For py-numpy itself, the section title for the defaults changed from DEFAULTS to ALL with
numpy-1.10 so that is configured accordingly. Also, for MKL to be used there must be a
section called, '[mkl]'. Finally, the mkl_rt library is needed, especially when built
with GCC. As such, that library is made explicit in the [mkl] section of the config file.

### CODE REVIEW

File: var/spack/repos/builtin/packages/mkl/package.py [hunk 1]
```diff
@@ -71,4 +71,17 @@ def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
         spack_env.set('MKLROOT', self.prefix)
 
     def setup_environment(self, spack_env, env):
+        # Remove paths that were guessed but are incorrect for this package.
+        env.remove_path('LIBRARY_PATH',
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mkl/package.py]: AFAIK Spack does not use `LD_LIBRARY_PATH` or alike anywhere, everything is RPATH'ed, why do you need all those changes?

File: var/spack/repos/builtin/packages/intel-parallel-studio/package.py [hunk 2]
```diff
@@ -86,7 +86,7 @@ def blas_libs(self):
         # TODO: TBB threading: ['libmkl_tbb_thread', 'libtbb', 'libstdc++']
         mkl_libs = find_libraries(
             mkl_integer + ['libmkl_core'] + mkl_threading,
-            root=join_path(self.prefix.lib, 'intel64'),
+            root=join_path(self.prefix, 'mkl', 'lib', 'intel64'),
```
Comment (davydden) [inline var/spack/repos/builtin/packages/intel-parallel-studio/package.py]: that looked like a bug, it's probably better have it in a separate commit.

File: var/spack/repos/builtin/packages/mkl/package.py [hunk 3]
```diff
@@ -71,4 +71,17 @@ def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
...
+        env.remove_path('LD_LIBRARY_PATH',
+                        join_path(self.prefix, 'lib'))
+        env.prepend_path('LD_LIBRARY_PATH',
+                         join_path(self.prefix, 'lib', 'intel64'))
+        env.prepend_path('LIBRARY_PATH',
+                         join_path(self.prefix, 'lib', 'intel64'))
+        env.prepend_path('LD_LIBRARY_PATH',
+                         join_path(self.prefix, 'lib', 'intel64'))
+        env.prepend_path('LIBRARY_PATH',
+                         join_path(self.prefix, 'lib', 'intel64'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mkl/package.py]: So you remove, and then prepend, and then prepend a second time?

File: var/spack/repos/builtin/packages/py-numpy/package.py [hunk 4]
```diff
@@ -76,11 +76,38 @@ def install(self, spec, prefix):
...
         if '+blas' in spec or '+lapack' in spec:
             with open('site.cfg', 'w') as f:
-                f.write('[DEFAULT]\n')
-                f.write('libraries=%s\n'    % ','.join(lapackblas.names))
+                if '^mkl' in spec:
+                    # Even though the intel-parallel-studio package provides
+                    # mkl that package will not work here with a non-Intel
+                    # compiler due to header collision.
+                    f.write('[mkl]\n')
+                    f.write('libraries=mkl_rt\n')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: i am not in favour of this change. `join(lapackblas.names)` depends on whole lot of variants (32/64 bits, threaded/serial, and alike). Once you use `mkl_rt` here it is now up to environment variable. In other words you introduce a way to use different BLAS/LAPACK in DAG. 

What was wrong exactly with `'join(lapackblas.names)`? At the end of the day we give the complete library lists there and it is bound to work. At least it does in q number of other packages without a need to use `mkl_rt` anywhere.

**Comment by davydden** [comment]
p.s. if `py-numpy` config system is crazy, so be it, but please try avoid using `mkl_rt` for consistency with other libraries and stick with `lapackblas.names`.

File: var/spack/repos/builtin/packages/intel-parallel-studio/package.py (hunk 2 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/intel-parallel-studio/package.py]: maybe in a separate PR which can be quickly merged. Then you will rebase this one to get rid of this change.

File: var/spack/repos/builtin/packages/mkl/package.py (hunk 3 above)
Comment (glennpj) [inline var/spack/repos/builtin/packages/mkl/package.py]: Oops, looks like duplicates sneaked in there. I will remove those.

**Comment by glennpj** [comment]
Okay; I guess I should have split this in multiple PRs. Regarding LD_LIBRARY_PATH, that is needed for the environment module not for the build.

Regarding the other comments, yes, the [mkl] section is needed in the numpy site.cfg, unless the config is removed it would fall back to MKLROOT. It seemed better to have the section in the config rather than rely on the environment variable.

Using mkl_rt is needed for numpy as far as I can tell. Intel states this as well here: https://software.intel.com/en-us/articles/numpyscipy-with-intel-mkl

> Building with GNU Compiler chain:
> 
> Make modifications to MKL section in the site.cfg as mentioned above. To build numpy and scipy with Gnu compilers, in the site.cfg file, you must link with mkl_rt only and any other linking method will not work.

Using mkl_rt can be directed by environment variables at run time but it does not depend on them.

I think that addresses the bulk of the comments. I am not really sure how to proceed here given the comments.

**Comment by davydden** [comment]
> Regarding LD_LIBRARY_PATH, that is needed for the environment module not for the build.

still don't understand why. Everything is built with RPATH's  in Spack, so `LD_LIBRARY_PATH` shall not be used at all. 

> Using mkl_rt is needed for numpy as far as I can tell.

I am not saying it does not build with `mkl_rt`, I am saying it is NOT the only way, actually it is even mentioned in SciPy's website:
https://www.scipy.org/scipylib/building/linux.html#intel-mkl-11-0-updated-dec-2012

So please try using `lapackblas.names`.

> Using mkl_rt can be directed by environment variables at run time but it does not depend on them.

it's not about that. The whole point of NOT using it is to provide consistent BLAS/LAPACK everwhere in the DAG. That is done through the extra class which holds libraries and have methods like `names()` and alike. 

> I am not really sure how to proceed here given the comments.

Please try (i) not to set LD_LIBRARY_PATH, it shall not be needed; (ii) use `lapackblas.names`, the link above suggests that it shall work.

**Comment by davydden** [comment]
FYI, from SciPy org (emphasize is mine):

> Instead of the layered linking approach for the Intel® MKL as shown above, you **MAY** also use the dynamic interface lib mkl_rt.lib.

**Comment by adamjstewart** [comment]
>> Regarding LD_LIBRARY_PATH, that is needed for the environment module not for the build.

> still don't understand why. Everything is built with RPATH's in Spack, so LD_LIBRARY_PATH shall not be used at all.

Python packages (like `py-scipy`) don't currently use RPATH. Also, anyone loading the MKL module outside of Spack for building their own code would surely want to set `LD_LIBRARY_PATH`. We do for all packages, it just so happens that the directory is wrong for MKL.

**Comment by davydden** [comment]
@adamjstewart :

> Python packages (like py-scipy) don't currently use RPATH. 

that explains a lot, thanks

File: var/spack/repos/builtin/packages/intel-parallel-studio/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/intel-parallel-studio/package.py]: I'm fine with this change being in this PR. You can't really test this PR without it I imagine.

Comment (davydden) [inline var/spack/repos/builtin/packages/intel-parallel-studio/package.py]: you surely can't. I am just saying to make things fast one could submit this change as a stand-alone PR, but still keep it here. Once a stand alone PR is merged, one can re-base this one on `develop`. 

But i don't really insist on this, just explain my rationale.

**Comment by glennpj** [comment]
@davydden Note the caveat that GCC requires mkl_rt, not the intel compiler, which is what your link refers to. Also, I did try to make it work with the standard blas_libs before agreeing with Intel that GCC requires mkl_rt.

Anyway, I will close this PR as clearly I have missed the mark.

**Comment by davydden** [comment]
> Note the caveat that GCC requires mkl_rt

but why? As I said above i had no problems building with MKL and GCC. Can you elaborate?

**Comment by davydden** [comment]
for the record, Intel's link advisor with **Select compiler: Gnu C/C++** says
```
-L${MKLROOT}/lib -Wl,-rpath,${MKLROOT}/lib -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl
```
so i see no reason why GCC would need `mkl_rt` for `py-numpy`.
